### ISSUE TITLE
Parallel Tests Failing

### ISSUE THREAD

**Issue by mathematicalmichael** [UT-CHG/BET#298]
@smattis they are not passing. clean clone of master branch, Python 2.7

Minimal output...

```
(py2) [BET]$ mpirun -np 2 nosetests
..................................SSSSSS......SS.......................S.S...S.S...............S.S.....S.S............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/contour.py:960: UserWarning: The following kwargs were not used by contour: 'triangles'
  s)
/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/contour.py:960: UserWarning: The following kwargs were not used by contour: 'triangles'
  s)
...E..................................................................................................................................................................................................................................................E.EEEE....................................EE.....................................E...E...............EE.................................................................................................................................................................................
======================================================================
ERROR: Test :met:`bet.sampling.adaptiveSampling.sampler.generalized_chains`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 455, in test_generalized_chains
    t_set, savefile, initial_sample_type, hot_start)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 146, in verify_samples
    hot_start=hot_start)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 492, in generalized_chains
    super(sampler, self).save(mdat, savefile, disc, globalize=False)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 141, in loadmat
    MR, file_opened = mat_reader_factory(file_name, appendmat, **kwargs)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 64, in mat_reader_factory
    byte_stream, file_opened = _open_file(file_name, appendmat)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 39, in _open_file
    return open(file_like, 'rb'), True
IOError: [Errno 2] No such file or directory: '11t11.mat'
-------------------- >> begin captured stdout << ---------------------
dim (1, 2)
COLD 0
dim (1, 2)
COLD then 0
1
50 5 None

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
root: INFO: COLD START
root: INFO: Current chain length: 2/10
root: INFO: Current chain length: 4/10
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
root: INFO: Current chain length: 10/10
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: COLD START
root: INFO: Current chain length: 2/5
root: INFO: Current chain length: 3/5
root: INFO: Current chain length: 4/5
root: INFO: Current chain length: 5/5
root: INFO: HOT START from partial run
root: INFO: HOT START using serial file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: HOT START from completed run
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Run :meth:`bet.sampling.adaptiveSampling.sampler.run_gen` and verify
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 305, in test_run_gen
    savefile)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 270, in run_gen
    initial_sample_type, criterion)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 422, in generalized_chains
    self.num_chains, criterion, globalize=False)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 470, in create_random_discretization
    savefile, globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 422, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1201, in save_discretization
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 272, in get_variables
    hdr, next_position = self.read_var_header()
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 231, in read_var_header
    raise TypeError('Expecting miMATRIX type here, got %d' % mdtype)
TypeError: Expecting miMATRIX type here, got 0
-------------------- >> begin captured logging << --------------------
root: INFO: COLD START
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Run :meth:`bet.sampling.adaptiveSampling.sampler.run_inc_dec` and verify
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 421, in test_run_inc_dec
    maximum, input_domain, t_set, savefile)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 366, in run_inc_dec
    t_set, savefile, initial_sample_type, criterion)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 270, in run_gen
    initial_sample_type, criterion)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 492, in generalized_chains
    super(sampler, self).save(mdat, savefile, disc, globalize=False)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1201, in save_discretization
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 272, in get_variables
    hdr, next_position = self.read_var_header()
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 219, in read_var_header
    raise ValueError("Did not read any bytes")
ValueError: Did not read any bytes
-------------------- >> begin captured logging << --------------------
root: INFO: COLD START
root: INFO: Current chain length: 2/10
root: INFO: Current chain length: 4/10
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
root: INFO: Current chain length: 10/10
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.user_samples`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 679, in test_compute_QoI_and_create_discretization
    input_sample_set, savefile)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 159, in verify_compute_QoI_and_create_discretization
    input_sample_set, savefile, globalize=True)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 422, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 272, in get_variables
    hdr, next_position = self.read_var_header()
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 231, in read_var_header
    raise TypeError('Expecting miMATRIX type here, got %d' % mdtype)
TypeError: Expecting miMATRIX type here, got 1668246623
-------------------- >> begin captured stdout << ---------------------
11t11 1
HERE HERE {'default_output_sample_set_values_local': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_local_index': array([[0, 1, 2, 3]]), 'default_output_sample_set_values': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_values_local': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_p_norm': array([[ 2.]]), '__header__': 'MATLAB 5.0 MAT-file Platform: posix, Created on: Tue Jan 29 17:14:31 2019', '__globals__': [], 'default_input_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_dim': array([[1]]), 'default_output_sample_set_p_norm': array([[ 2.]]), '__version__': '1.0', 'default_input_sample_set_values': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_dim': array([[1]]), 'num_samples': array([[4]])} 4
1t1 1
HERE HERE {'default_output_sample_set_values_local': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_local_index': array([[0, 1, 2, 3]]), 'default_output_sample_set_values': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_values_local': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_p_norm': array([[ 2.]]), '__header__': 'MATLAB 5.0 MAT-file Platform: posix, Created on: Tue Jan 29 17:14:31 2019', '__globals__': [], 'default_input_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_dim': array([[1]]), 'default_output_sample_set_p_norm': array([[ 2.]]), '__version__': '1.0', 'default_input_sample_set_values': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_dim': array([[1]]), 'num_samples': array([[4]])} 4
3to1 3

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.create_random_discretization`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 792, in test_create_random_discretization
    savefile)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 242, in verify_create_random_discretization
    saved_disc = bet.sample.load_discretization(savefile)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1346, in load_discretization
    discretization_name+attrname))
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 136, in load_sample_set
    mdat = sio.loadmat(file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 141, in loadmat
    MR, file_opened = mat_reader_factory(file_name, appendmat, **kwargs)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 64, in mat_reader_factory
    byte_stream, file_opened = _open_file(file_name, appendmat)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 39, in _open_file
    return open(file_like, 'rb'), True
IOError: [Errno 2] No such file or directory: '3to2.mat'
-------------------- >> begin captured logging << --------------------
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Tests :meth:`bet.sampling.basicSampling.loadmat`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 46, in test_loadmat
    (os.path.join(local_path, 'testfile1')), globalize=True)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 272, in get_variables
    hdr, next_position = self.read_var_header()
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 219, in read_var_header
    raise ValueError("Did not read any bytes")
ValueError: Did not read any bytes

----------------------------------------------------------------------
Ran 648 tests in 143.493s

FAILED (SKIP=8, errors=6)
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.
-------------------------------------------------------
............................................................................................
======================================================================
ERROR: Test :meth:`bet.postProcess.plotDomains.show_data_domain_multi`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_postProcess/test_plotDomains.py", line 331, in test_show_data_domain_multi
    self.check_show_data_domain_multi(rm, rc, qn, sd)
  File "/home/mpilosov/Packages/BET/test/test_postProcess/test_plotDomains.py", line 345, in check_show_data_domain_multi
    ref_colors=ref_colors, showdim=showdim)
  File "/home/mpilosov/Packages/BET/bet/postProcess/plotDomains.py", line 671, in show_data_domain_multi
    filenames=filenames, file_extension=file_extension)
  File "/home/mpilosov/Packages/BET/bet/postProcess/plotDomains.py", line 762, in show_data_domain_2D
    pad_inches=.2)
  File "/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/pyplot.py", line 710, in savefig
    res = fig.savefig(*args, **kwargs)
  File "/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/figure.py", line 2035, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 2263, in print_figure
    **kwargs)
  File "/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/backends/backend_agg.py", line 526, in print_png
    with cbook.open_file_cm(filename_or_obj, "wb") as fh:
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/cbook/__init__.py", line 624, in open_file_cm
    fh, opened = to_filehandle(path_or_file, mode, True, encoding)
  File "/home/mpilosov/.local/lib/python2.7/site-packages/matplotlib/cbook/__init__.py", line 609, in to_filehandle
    fh = io.open(fname, flag, encoding=encoding)
IOError: [Errno 2] No such file or directory: 'figs/domain_q3_q4.png'

======================================================================
ERROR: Test :met:`bet.sampling.adaptiveSampling.sampler.generalized_chains`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 455, in test_generalized_chains
    t_set, savefile, initial_sample_type, hot_start)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 146, in verify_samples
    hot_start=hot_start)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 443, in generalized_chains
    num_chains=self.num_chains)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 72, in loadmat
    disc = sample.load_discretization(save_file)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1346, in load_discretization
    discretization_name+attrname))
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 136, in load_sample_set
    mdat = sio.loadmat(file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 141, in loadmat
    MR, file_opened = mat_reader_factory(file_name, appendmat, **kwargs)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 65, in mat_reader_factory
    mjv, mnv = get_matfile_version(byte_stream)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/miobase.py", line 224, in get_matfile_version
    raise MatReadError("Mat file appears to be empty")
MatReadError: Mat file appears to be empty
-------------------- >> begin captured stdout << ---------------------
dim (1, 2)
COLD 0
dim (1, 2)
COLD then 0
1
50 5 None
HOT 0
dim (1, 2)
COLD then 0
2
50 5 None
HOT 0
dim (1, 2)
COLD 0
dim (1, 2)
COLD then 0
1

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
root: INFO: COLD START
root: INFO: Current chain length: 2/10
root: INFO: Current chain length: 4/10
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
root: INFO: Current chain length: 10/10
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: COLD START
root: INFO: Current chain length: 2/5
root: INFO: Current chain length: 3/5
root: INFO: Current chain length: 4/5
root: INFO: Current chain length: 5/5
root: INFO: HOT START from partial run
root: INFO: HOT START using serial file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: HOT START from completed run
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
root: INFO: Current chain length: 10/10
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: COLD START
root: INFO: Current chain length: 2/5
root: INFO: Current chain length: 3/5
root: INFO: Current chain length: 4/5
root: INFO: Current chain length: 5/5
root: INFO: HOT START from completed run
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
root: INFO: Current chain length: 10/10
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: COLD START
root: INFO: Current chain length: 2/10
root: INFO: Current chain length: 4/10
root: INFO: Current chain length: 6/10
root: INFO: Current chain length: 8/10
root: INFO: Current chain length: 10/10
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: COLD START
root: INFO: Current chain length: 2/5
root: INFO: Current chain length: 3/5
root: INFO: Current chain length: 4/5
root: INFO: Current chain length: 5/5
root: INFO: HOT START from partial run
root: INFO: HOT START using serial file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Run :meth:`bet.sampling.adaptiveSampling.sampler.run_inc_dec` and verify
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 421, in test_run_inc_dec
    maximum, input_domain, t_set, savefile)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 366, in run_inc_dec
    t_set, savefile, initial_sample_type, criterion)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 270, in run_gen
    initial_sample_type, criterion)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 422, in generalized_chains
    self.num_chains, criterion, globalize=False)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 470, in create_random_discretization
    savefile, globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 422, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 141, in loadmat
    MR, file_opened = mat_reader_factory(file_name, appendmat, **kwargs)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 65, in mat_reader_factory
    mjv, mnv = get_matfile_version(byte_stream)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/miobase.py", line 224, in get_matfile_version
    raise MatReadError("Mat file appears to be empty")
MatReadError: Mat file appears to be empty
-------------------- >> begin captured logging << --------------------
root: INFO: COLD START
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Run :meth:`bet.sampling.adaptiveSampling.sampler.run_tk` and verify
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_adaptiveSampling.py", line 363, in test_run_tk
    maximum, input_domain, kernel_rD, savefile)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 321, in run_tk
    initial_sample_type, criterion)
  File "/home/mpilosov/Packages/BET/bet/sampling/adaptiveSampling.py", line 422, in generalized_chains
    self.num_chains, criterion, globalize=False)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 470, in create_random_discretization
    savefile, globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 422, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 272, in get_variables
    hdr, next_position = self.read_var_header()
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 231, in read_var_header
    raise TypeError('Expecting miMATRIX type here, got %d' % mdtype)
TypeError: Expecting miMATRIX type here, got 0
-------------------- >> begin captured logging << --------------------
root: INFO: COLD START
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Test :meth:`bet.sampling.basicSampling.sampler.user_samples`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 679, in test_compute_QoI_and_create_discretization
    input_sample_set, savefile)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 159, in verify_compute_QoI_and_create_discretization
    input_sample_set, savefile, globalize=True)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 422, in compute_QoI_and_create_discretization
    self.save(mdat, savefile, discretization, globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sampling/basicSampling.py", line 279, in save
    globalize=globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 272, in get_variables
    hdr, next_position = self.read_var_header()
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio5.py", line 219, in read_var_header
    raise ValueError("Did not read any bytes")
ValueError: Did not read any bytes
-------------------- >> begin captured stdout << ---------------------
11t11 1
HERE HERE {'default_output_sample_set_values_local': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_local_index': array([[0, 1, 2, 3]]), 'default_output_sample_set_values': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_values_local': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_p_norm': array([[ 2.]]), '__header__': 'MATLAB 5.0 MAT-file Platform: posix, Created on: Tue Jan 29 17:14:30 2019', '__globals__': [], 'default_input_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_dim': array([[1]]), 'default_output_sample_set_p_norm': array([[ 2.]]), '__version__': '1.0', 'default_input_sample_set_values': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_dim': array([[1]]), 'num_samples': array([[4]])} 4
1t1 1
HERE HERE {'default_output_sample_set_values_local': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_local_index': array([[0, 1, 2, 3]]), 'default_output_sample_set_values': array([[ 0.84147098],
       [ 0.84147098],
       [ 0.84147098],
       [ 0.84147098]]), 'default_input_sample_set_values_local': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_p_norm': array([[ 2.]]), '__header__': 'MATLAB 5.0 MAT-file Platform: posix, Created on: Tue Jan 29 17:14:30 2019', '__globals__': [], 'default_input_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_dim': array([[1]]), 'default_output_sample_set_p_norm': array([[ 2.]]), '__version__': '1.0', 'default_input_sample_set_values': array([[ 1.],
       [ 1.],
       [ 1.],
       [ 1.]]), 'default_output_sample_set_dim': array([[1]]), 'num_samples': array([[4]])} 4
3to1 3
HERE HERE {'default_output_sample_set_values_local': array([[ 3.],
       [ 3.],
       [ 3.],
       [ 3.]]), 'default_input_sample_set_local_index': array([[0, 1, 2, 3]]), 'default_output_sample_set_values': array([[ 3.],
       [ 3.],
       [ 3.],
       [ 3.]]), 'default_input_sample_set_values_local': array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.]]), 'default_output_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_p_norm': array([[ 2.]]), '__header__': 'MATLAB 5.0 MAT-file Platform: posix, Created on: Tue Jan 29 17:14:30 2019', '__globals__': [], 'default_input_sample_set_sample_set_type': array([u'bet.sample.sample_set'],
      dtype='<U21'), 'default_input_sample_set_dim': array([[3]]), 'default_output_sample_set_p_norm': array([[ 2.]]), '__version__': '1.0', 'default_input_sample_set_values': array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.]]), 'default_output_sample_set_dim': array([[1]]), 'num_samples': array([[4]])} 4
3to2 3

--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
root: INFO: No sample_set named default_emulated_input_sample_set with _dim in file
root: INFO: No sample_set named default_emulated_output_sample_set with _dim in file
root: INFO: No sample_set named default_output_probability_set with _dim in file
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Tests :class:`bet.sampling.basicSampling.sampler.loadmat`.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/nose-1.3.7-py2.7.egg/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/mpilosov/Packages/BET/test/test_sampling/test_basicSampling.py", line 113, in test_loadmat_parallel
    file_name1, globalize=False)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 1195, in save_discretization
    discretization_name+attrname, globalize)
  File "/home/mpilosov/Packages/BET/bet/sample.py", line 82, in save_sample_set
    new_mdat = sio.loadmat(local_file_name)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio.py", line 142, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio4.py", line 399, in get_variables
    mdict[name] = self.read_var_array(hdr)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio4.py", line 374, in read_var_array
    return self._matrix_reader.array_from_header(header, process)
  File "/home/mpilosov/anaconda3/envs/py2/lib/python2.7/site-packages/scipy/io/matlab/mio4.py", line 146, in array_from_header
    raise TypeError('No reader for class code %s' % mclass)
TypeError: No reader for class code 4

----------------------------------------------------------------------
Ran 648 tests in 144.250s

FAILED (SKIP=8, errors=6)
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[55423,1],1]
  Exit code:    1
--------------------------------------------------------------------------


```

**Comment by mathematicalmichael** [comment]
basically, it seems that none of the sampling module tests work...

**Comment by mathematicalmichael** [comment]
source of error is that the decorator class used to skip the test is being ignored.
Why? 
Well, `bet.Comm` is reporting `.size` of 1 even with `NPROCS` = 2.

**Comment by mathematicalmichael** [comment]
Thanks, @smattis. That's helpful. I actually started investigating this morning (hence the comment and self-assigning), and literally just discovered problems trying to link MPI. 

It's frustrating that my testing environment (the docker containers) are finicky like this, but I was also getting problems with "could not find files" on my workstation, which is realllly similar to the container environment as far as software... So strange. 

Here is the primary difference as far as I can tell..
Travis CI is on Ubuntu 14 and I'm on 18.

**Comment by mathematicalmichael** [comment]
I'd like to point out that this container environment is definitely where the future is trending towards (due to ease of set-up for new users, reproducibility through tools like binder, etc), and I would like to prioritize testing on modern Ubuntu releases since the security is greatly improved.

**Comment by smattis** [comment]
Can you post the YAML file that you are using to setup your environment?

**Comment by mathematicalmichael** [comment]
I'm not using one.
```
RUN conda create --quiet --yes -p $CONDA_DIR/envs/python2 python=2.7 ipython ipykernel kernda     numpy pandas matplotlib ipywidgets yaml
```

is the line in the Dockerfile that does the work.

But hold on... The issue here is that in my environment, _something_ is causing `Comm` to report the wrong number of processors in parallel. The CORRECT output for `mpirun -np 2 test_plotDomains.py` should be "SKIPPED" ! (AKA the decorator on top of the test is being ignored... this isn't quite an environment issue perhaps...). Let me try another logic (non-decorator) to skip the test in parallel.

**Comment by mathematicalmichael** [comment]
```
-@unittest.skipIf(comm.size > 1, 'Only run in serial')
+#@unittest.skipIf(comm.size > 1, 'Only run in serial')
 class test_plotDomains(unittest.TestCase):
     """
     Test :meth:`bet.postProcess.plotP.calculate_1D_marginal_probs` and
@@ -33,6 +33,8 @@ class test_plotDomains(unittest.TestCase):
         """
         Set up problem.
         """
+        if comm.size > 1:
+            self.skipTest('Test skipped in parallel mode.')
```

This logic seems to fix the issue.
I'll go ahead and push this commit through since it's completely equivalent and seems more robust

**Comment by smattis** [comment]
Ok, it's weird that the decorator is being skipped, though. This is equivalent though.

Please do a pull request, and let me review and merge rather than pushing directly to master!

**Comment by mathematicalmichael** [comment]
certainly is weird that the decorator is being skipped, I agree. 
I do think this is the only place it shows up though.

**Comment by smattis** [comment]
@mathematicalmichael Sorry, but I removed your write privileges because of that. You should always make changes via a pull request and make sure all of the checks pass. That push triggered a CI build on Travis, so if it passes I'm ok with not going through the hassle of reverting and doing a pull request. 

For now, just do everything on your fork and do a pull request. I should prep you on our protocols before giving you write access.

**Comment by mathematicalmichael** [comment]
Looking at the output of the error messages above, and re-running nosetests as a whole actually led me to find a bunch of other errors as well in this environment. I don't know what is happening but I'll meet with Troy and see if he's able to pass parallel tests on his machines. I'll also try an Ubuntu 14 Docker container just to see if that could possibly be part of this... 

And yes please do that. Monty's group was the only other one I've worked with, and he just has everyone push everything to master. My sincerest apologies. I'm totally fine with the PR approach, I just want to be able to label issues for organization purposes.

**Comment by mathematicalmichael** [comment]
Here is how I've set up the environment (on top of docker-stacks/datascience-notebook), which builds on top of Ubuntu 18.

https://github.com/mathematicalmichael/jupyterhub-deploy-docker/blob/master/singleuser/Dockerfile

**Comment by mathematicalmichael** [comment]
I have tried `conda install mpi4py` and still ran into these problems. 
I tried again on a fresh Ubuntu 18 machine and new conda env. Same parallel problems. Plenty of failed tests having nothing to do with `plotDomains` (unless... that decorator shows up in other places and is indeed the source of all the errors... we'll see). 

I _also_ tried again on my Macbook Pro (Mohave), fresh install of `mpi4py`, with a fresh clone of `master` and there too I could not get the tests to pass in parallel. 

What environment are you testing in that is allowing you to pass these parallel tests?

**Comment by smattis** [comment]
Try installing mpi4py with pip instead of conda.

**Comment by mathematicalmichael** [comment]
perhaps it's how I'm setting up parallelism? Removed conda install of mpi4py, then

```
pip install ipyparallel mpi4py
# Didn't work. Looked online. The following is supposed to help. Changes errors but still failing.
ipython profile create --parallel --profile=mpi
```

**Comment by smattis** [comment]
It might be something like that. I'm wondering if it is something to do with your MPI implementation. I don't see you building mpich2 libraries anywhere in your Dockerfile, so I don't know where they are being built. I am guessing you don't have a working MPI implementation, though.  Maybe add
`sudo apt-get install mpich2 libmpich2-dev`
to your Dockerfile

**Comment by mathematicalmichael** [comment]
are those the right ones? trying on my workstation first since it is similar to the docker containers (the original thread was from there, and there are only 6 errors vs the 30 in my docker container... MPI set up is definitely the issue...) 

```
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package libmpich2-dev is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  libmpich-dev:i386 libmpich-dev

Package mpich2 is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source
However the following packages replace it:
  mpich:i386 mpich

E: Package 'mpich2' has no installation candidate
E: Package 'libmpich2-dev' has no installation candidate
```

Tried removing the 2 and the install gets immediately aborted...

**Comment by mathematicalmichael** [comment]
taking abnormally long to run parallel tests. Something's hung up...

Inspecting the installation output from your suggestion, I see
```
update-alternatives: using /usr/bin/mpirun.mpich to provide /usr/bin/mpirun (mpirun) in auto mode
update-alternatives: warning: skip creation of /usr/share/man/man1/mpirun.1.gz because associated file /usr/share/man/man1/mpirun.mpich.1.gz (of link group mpirun) doesn't exist
update-alternatives: warning: skip creation of /usr/share/man/man1/mpiexec.1.gz because associated[~]#
```

Okay, so linking is not being done correctly? Was dropping the trailing `2` a problem?

**Comment by mathematicalmichael** [comment]
You were on to it... In the containers...  
Bless your soul. I owe you a beer. 
Go ahead an revert my changes, I want you to review the set of commits in the PR for the upgrade. This isn't how I wanted it to go... But at least it helped diagnose the problem I was having. I will make sure to include mpi installation instructions in the README

```
(python2) [BET]$ mpirun -np 1 nosetests
.................SSS...S............S..S........S...S............................................................................................................................................................................................................................................................................................../opt/conda/envs/python2/lib/python2.7/site-packages/matplotlib/contour.py:1004: UserWarning: The following kwargs were not used by contour: 'triangles'
  s)
........../opt/conda/envs/python2/lib/python2.7/site-packages/scipy-1.1.0-py2.7-linux-x86_64.egg/scipy/fftpack/basic.py:160: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  z[index] = x
......................../opt/conda/envs/python2/lib/python2.7/site-packages/numpy/core/fromnumeric.py:83: RuntimeWarning: invalid value encountered in reduce
  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
...................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 648 tests in 144.146s

OK (SKIP=8)
```

**Comment by smattis** [comment]
Ok. I'm glad it's working. So you just ended up building mpich and its libraries, and it worked. Is that right? I guess mpich2 is just called mpich in Ubuntu 18?

I reverted the commits for your bad pull request in a new pull request and will merge it once the checks pass. Then you should do a new pull request fixing the issue of the tests not being skipped in parallel (or maybe your new mpi build fixes that?)

**Comment by mathematicalmichael** [comment]
Yes, indeed! That is what fixed it, and that must be what it is called. 

Now that MPI is fixed, I'll make sure to double-check the decorator. I suspect I'll end up reverting that commit. Still interesting though that it allowed one of the failing tests to pass despite the logic being identical.... weird. If it does indeed make an impact, I'll submit a PR. 

When I eventually push the upgrade... should I specify a higher-versioned OS for Travis? Or does that not really matter technically speaking?

**Comment by smattis** [comment]
Right now, it's using the default linux OS, and I would prefer to just keep doing that. Changes in linux version should not have much of an effect.

**Comment by mathematicalmichael** [comment]
got it. 

okay. closing out this issue.

Since I couldn't get the commit history messages to move over properly after messing up, I'm going to re-apply changes one-by-one to have a clean PR and do this correctly.
